File: tests/test_parser.cpp
/***************************************************************************
 *            tests/test_parser.cpp
 *
 *  Copyright  2021  Mirco De Marchi
 *
 ****************************************************************************/

/*
 *  This file is part of Ariadne.
 *
 *  Ariadne is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  Ariadne is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with Ariadne.  If not, see <https://www.gnu.org/licenses/>.
 */

#include "test.hpp"
#include "parser/parser.hpp"

#include <vector>
#include <map>
#include <string>


using namespace std;
using namespace Ariadne;

class TestParser {
public:
    void test() {
        ARIADNE_TEST_CALL(test_parse());
        ARIADNE_TEST_CALL(test_is());
        ARIADNE_TEST_CALL(test_convert());
    }

private:
    void test_parse() {
        auto parser = Parser();
        auto test_vec = std::map<std::string, ParserType>{
            { "1.2",           ParserType::FLOAT  },
            { "+0.0",          ParserType::FLOAT  },
            { "-0.0",          ParserType::FLOAT  },
            { "+1e-10",        ParserType::FLOAT  },
            { "true",          ParserType::BOOL   },
            { "1",             ParserType::INT    },
            { "-1",            ParserType::INT    },
            { "+0",            ParserType::INT    },
            { "-0",            ParserType::INT    },
            { "\"string\"",    ParserType::STRING },
            { "123ariadne456", ParserType::STRING },
            { std::string{},   ParserType::NONE   },
        };

        auto keys = std::vector<std::string>{};
        auto values = std::vector<ParserType>{};
        for (const auto& [key, value]: test_vec)
        {
            keys.push_back(key);
            values.push_back(value);
        }

        auto parsed_values = parser(keys);
        ARIADNE_TEST_ASSERT(parsed_values == values);
        ARIADNE_TEST_PRINT(parsed_values);

        auto values_cpy = std::vector<ParserType>(values);
        values_cpy.push_back(ParserType::NONE);
        ARIADNE_TEST_ASSERT(parsed_values != values_cpy);
    }

    void test_is() {
        ARIADNE_TEST_ASSERT(Parser::is_float("-0.3"));
        ARIADNE_TEST_ASSERT(Parser::is_float(".245"));
        ARIADNE_TEST_ASSERT(Parser::is_bool("false"));
        ARIADNE_TEST_ASSERT(Parser::is_int("1234"));
        ARIADNE_TEST_ASSERT(Parser::is_string("ariadne123dl"));
    }

    void test_convert() {
        float f;
        ARIADNE_TEST_ASSERT(convert("1.2", &f));
        ARIADNE_TEST_WITHIN(f, 1.2, 0.0000001);
        bool b;
        ARIADNE_TEST_ASSERT(convert("true", &b));
        ARIADNE_TEST_EQUAL(b, true);
        int i;
        ARIADNE_TEST_ASSERT(convert("1", &i));
        ARIADNE_TEST_EQUAL(i, 1);
        std::string s;
        ARIADNE_TEST_ASSERT(convert("1", &s));
        ARIADNE_TEST_EQUAL(s, "1");

        ARIADNE_TEST_ASSERT(!convert("1string", &i));
    }
};

int main() {
    TestParser().test();
    return ARIADNE_TEST_FAILURES;
}

File: python/tests/test_parser.py
#!/usr/bin/python3

##############################################################################
#            test_parser.py
#
#  Copyright  2006-20  Mirco De Marchi
##############################################################################

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

import os
from pyedgelearning.parser import *


def test_parser():
    assert(True)
    return # Working test with all datasets in data folder.
    resource_path = os.path.join(os.path.dirname(__file__), "..", "..", "data")
    train_mnist, test_mnist = load_mnist(resource_path)

    assert(train_mnist.size() == 60000)
    assert(train_mnist.feature_size() == (28 * 28 + 10))
    assert(train_mnist.sequence_size == 1)
    assert(train_mnist.label_idx == list(range(28 * 28, 28 * 28 + 10)))

    assert(test_mnist.size() == 10000)
    assert(test_mnist.feature_size() == (28 * 28 + 10))
    assert(test_mnist.sequence_size == 1)
    assert(test_mnist.label_idx == list(range(28 * 28, 28 * 28 + 10)))

    train_cifar10, test_cifar10 = load_cifar10(resource_path)

    assert(train_cifar10.size() == 50000)
    assert(train_cifar10.feature_size() == (32 * 32 * 3 + 10))
    assert(train_cifar10.sequence_size == 1)
    assert(train_cifar10.label_idx == list(range(32 * 32 * 3, 32 * 32 * 3 + 10)))

    assert(test_cifar10.size() == 10000)
    assert(test_cifar10.feature_size() == (32 * 32 * 3 + 10))
    assert(test_cifar10.sequence_size == 1)
    assert(test_cifar10.label_idx == list(range(32 * 32 * 3, 32 * 32 * 3 + 10)))

    train_cifar100, test_cifar100 = load_cifar100(resource_path)

    assert(train_cifar100.size() == 10000)
    assert(train_cifar100.feature_size() == (32 * 32 * 3 + 100))
    assert(train_cifar100.sequence_size == 1)
    assert(train_cifar100.label_idx == list(range(32 * 32 * 3, 32 * 32 * 3 + 100)))

    assert(test_cifar100.size() == 10000)
    assert(test_cifar100.feature_size() == (32 * 32 * 3 + 100))
    assert(test_cifar100.sequence_size == 1)
    assert(test_cifar100.label_idx == list(range(32 * 32 * 3, 32 * 32 * 3 + 100)))

